api/src/main/java/org/apache/iceberg/transforms/Timestamps.java

```diff
@@ -32,24 +32,78 @@
 import org.apache.iceberg.util.SerializableFunction;
 
 enum Timestamps implements Transform<Long, Integer> {
-  MICROS_TO_YEAR(ChronoUnit.YEARS, "year", MicrosToYears.INSTANCE),
-  MICROS_TO_MONTH(ChronoUnit.MONTHS, "month", MicrosToMonths.INSTANCE),
-  MICROS_TO_DAY(ChronoUnit.DAYS, "day", MicrosToDays.INSTANCE),
-  MICROS_TO_HOUR(ChronoUnit.HOURS, "hour", MicrosToHours.INSTANCE),
+  MICROS_TO_YEAR(ChronoUnit.YEARS, "year", TimestampUnit.MICROS),
+  MICROS_TO_MONTH(ChronoUnit.MONTHS, "month", TimestampUnit.MICROS),
+  MICROS_TO_DAY(ChronoUnit.DAYS, "day", TimestampUnit.MICROS),
+  MICROS_TO_HOUR(ChronoUnit.HOURS, "hour", TimestampUnit.MICROS),
 
-  NANOS_TO_YEAR(ChronoUnit.YEARS, "year", NanosToYears.INSTANCE),
-  NANOS_TO_MONTH(ChronoUnit.MONTHS, "month", NanosToMonths.INSTANCE),
-  NANOS_TO_DAY(ChronoUnit.DAYS, "day", NanosToDays.INSTANCE),
-  NANOS_TO_HOUR(ChronoUnit.HOURS, "hour", NanosToHours.INSTANCE);
+  NANOS_TO_YEAR(ChronoUnit.YEARS, "year", TimestampUnit.NANOS),
+  NANOS_TO_MONTH(ChronoUnit.MONTHS, "month", TimestampUnit.NANOS),
+  NANOS_TO_DAY(ChronoUnit.DAYS, "day", TimestampUnit.NANOS),
+  NANOS_TO_HOUR(ChronoUnit.HOURS, "hour", TimestampUnit.NANOS);
+
+  @Immutable
+  static class Apply implements SerializableFunction<Long, Integer> {
+    private final ChronoUnit granularity;
+    private final TimestampUnit timestampUnit;
+
+    Apply(ChronoUnit granularity, TimestampUnit timestampUnit) {
+      this.granularity = granularity;
+      this.timestampUnit = timestampUnit;
+    }
+
+    @Override
+    public Integer apply(Long timestamp) {
+      if (timestamp == null) {
+        return null;
+      }
+
+      switch (timestampUnit) {
+        case MICROS:
+          switch (granularity) {
+            case YEARS:
+              return DateTimeUtil.microsToYears(timestamp);
+            case MONTHS:
+              return DateTimeUtil.microsToMonths(timestamp);
+            case DAYS:
+              return DateTimeUtil.microsToDays(timestamp);
+            case HOURS:
+              return DateTimeUtil.microsToHours(timestamp);
+            default:
+              throw new UnsupportedOperationException("Unsupported time unit: " + granularity);
+          }
```

Comment:
nit: add new lines to separate the block
https://iceberg.apache.org/contribute/?h=block+spacing#block-spacing